test/docker-e2e/e2e_pfm_test.go

```diff
@@ -0,0 +1,374 @@
+		}
+
+		finalConC := s.getBalance(ctx, s.chainC, walletC.GetFormattedAddress(), denomOnC)
+		finalBonB := s.getBalance(ctx, s.chainB, walletB.GetFormattedAddress(), denomOnB)
+		finalA := s.getBalance(ctx, s.chainA, walletA.GetFormattedAddress(), baseDenom)
+
+		conditionsMet := true
+		if !finalConC.Sub(initialConC).Equal(amt) {
+			s.T().Logf("Chain C has not received tokens yet (got %s, expected %s)", finalConC.Sub(initialConC).String(), amt.String())
+			conditionsMet = false
+		} else {
+			s.T().Logf("Chain C received correct amount")
+		}
+
+		if !finalBonB.Equal(initialBonB) {
+			s.T().Logf("Chain B incorrectly retained funds (got %s, expected %s)", finalBonB.String(), initialBonB.String())
+			conditionsMet = false
+		} else {
+			s.T().Logf("Chain B correctly forwarded (balance unchanged)")
+		}
+
+		if !finalA.Equal(expectedAAfter) {
+			s.T().Logf("Chain A balance incorrect (got %s, expected %s)", finalA.String(), expectedAAfter.String())
+			conditionsMet = false
+		} else {
+			s.T().Logf("Chain A balance correct")
+		}
+
+		return conditionsMet, nil
+	})
+
+	s.Require().NoError(err, "PFM multi-hop transfer failed")
+
+	s.T().Logf("Verifying PFM used two-hop path (not direct A->C)")
+
+	twoHopDenom := calculateIBCDenomTrace(
+		[]struct{ port, channel string }{
+			{s.chAToB.CounterpartyPort, s.chAToB.CounterpartyID},
+			{s.chBToC.CounterpartyPort, s.chBToC.CounterpartyID},
+		}, baseDenom,
+	)
+	twoHopBalance := s.getBalance(ctx, s.chainC, walletC.GetFormattedAddress(), twoHopDenom)
+	s.T().Logf("Two-hop path (A->B->C) balance: %s %s", twoHopBalance.String(), twoHopDenom)
+
+	s.assertOnlyTwoHopDenom(ctx, s.chainC, walletC.GetFormattedAddress(), twoHopDenom)
+
+	if twoHopBalance.Equal(sdkmath.NewInt(sendAmount)) {
+		s.T().Logf("PFM used two-hop path (A->B->C)")
+	} else {
+		s.T().Fatalf("expected %d via two-hop path, got %s", sendAmount, twoHopBalance.String())
+	}
+
+	// Final verification
+	finalBonB := s.getBalance(ctx, s.chainB, walletB.GetFormattedAddress(), denomOnB)
+	s.Require().True(finalBonB.Equal(initialBonB), "chain-b retained funds: before=%s after=%s", initialBonB.String(), finalBonB.String())
+
+	finalABalance := s.getBalance(ctx, s.chainA, walletA.GetFormattedAddress(), baseDenom)
+	s.Require().True(expectedAAfter.Equal(finalABalance), "chain-a balance mismatch: expected %s got %s", expectedAAfter.String(), finalABalance.String())
+}
+
+func (s *PFMTestSuite) TearDownTest() {
+	ctx := context.Background()
+
+	chains := []*cosmos.Chain{s.chainA, s.chainB, s.chainC}
+	for _, chain := range chains {
+		if chain == nil {
+			continue
+		}
+		if err := chain.Remove(ctx); err != nil {
+			s.T().Logf("Error stopping chain %s: %v", chain.GetChainID(), err)
+		}
+	}
+
+	if s.hermes != nil {
+		if err := s.hermes.Remove(ctx); err != nil {
+			s.T().Logf("Error stopping hermes: %v", err)
+		}
+	}
+}
+
+func (s *PFMTestSuite) getAllBalances(ctx context.Context, chain *cosmos.Chain, address string) sdk.Coins {
```

Comment:
I think we can use the built in tastora fn to do this https://github.com/celestiaorg/tastora/blob/1a4acde331d5fb23bc9284b4a8afe79839868909/framework/testutil/query/query.go#L12